onnxruntime/core/session/custom_ops.cc

```diff
@@ -755,66 +757,89 @@ KernelCreateInfo CreateKernelCreateInfo(const std::string& domain, const OrtCust
+  ORT_ENFORCE(ops.size() > 0, "No kernels to registers.");
   int undefined = 0;
 
+  const OrtCustomOp* op = *ops.begin();
   ONNX_NAMESPACE::OpSchema schema(op->GetName(op), "custom op registered at runtime", 0);
 
-  for (size_t i = 0; i < input_count; i++) {
+  auto create_type_constraint = [&ops, &schema, &undefined](const OrtCustomOp* op, int count, int i, bool is_input) {
     onnx::OpSchema::FormalParameterOption option = onnx::OpSchema::FormalParameterOption::Single;
     bool is_homogeneous = true;
     int min_arity = 1;
 
     // The OrtCustomOp interface did not support the methods to query input/output characteristics before
     // ORT API version 8. So, query the relevant methods ONLY from API version 8 onwards.
     if (op->version >= min_ort_version_with_optional_io_support) {
-      const auto characteristic = op->GetInputCharacteristic(op, i);
+      const auto characteristic = is_input ? op->GetInputCharacteristic(op, i) : op->GetOutputCharacteristic(op, i);
 
       // Support for optional and variadic inputs/output was added in versions 8 and 14, respectively.
       if (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
         option = onnx::OpSchema::FormalParameterOption::Optional;
       } else if ((op->version >= min_ort_version_with_variadic_io_support) &&
                  (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC)) {
-        ORT_ENFORCE(i == input_count - 1, "Only the last input to a custom op may be marked variadic.");
+        ORT_ENFORCE(i == count - 1, "Only the last ", (is_input ? "input" : "output"),
+                    " to a custom op may be marked variadic.");
         option = onnx::OpSchema::FormalParameterOption::Variadic;
-        min_arity = op->GetVariadicInputMinArity(op);
-        is_homogeneous = static_cast<bool>(op->GetVariadicInputHomogeneity(op));
+        min_arity = is_input ? op->GetVariadicInputMinArity(op) : op->GetVariadicOutputMinArity(op);
+        is_homogeneous = static_cast<bool>(is_input
+                                               ? op->GetVariadicInputHomogeneity(op)
+                                               : op->GetVariadicOutputHomogeneity(op));
       }
     }
 
-    const auto type = op->GetInputType(op, i);
-    if (type == ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED) {
-      undefined++;
+    std::unordered_set<ONNXTensorElementDataType> all_types;
+    for (auto o : ops) {
+      ORT_ENFORCE(i < (is_input ? o->GetInputTypeCount(o) : o->GetOutputTypeCount(o)),
+                  "Another version of operator '", schema.Name(),
+                  "'has less ", (is_input ? "inputs" : "outputs"),
+                  ". onnxruntime allows the overloading of an operator "
+                  "if all versions have the same number of declared ",
+                  (is_input ? "inputs" : "outputs"), ".");
```

Comment:
If they need to be the same, does this need to also detect if a later registration has less inputs? 

If so, could we instead loop before calling create_type_constraint and check `input_count`/`output_count` matches instead of checking every value of `i`?

Reply:
Once the schema is created, the function CreateCustomRegistry checks all overlaoded operators are compatible with the created schema (with method IsCompatible). This verification is handled there (here: https://github.com/microsoft/onnxruntime/pull/17497/files#diff-eb59007a97532913933bd6d031e8f441215420a2e73657094e85c17a52976d17R1054). In that loop, I just check that one index is not out of boundary because the verification I just mentioned happens after this. That's why I did not add another test. I can move this verification earlier if you think it is better.

Reply:
I'd prefer to avoid doing partial verification in multiple places. 

If IsCompatible enforces the input/output counts match, I'd simply use that and add a comment here saying that IsCompatible is responsible for the validation (so the next developer doesn't have to search for it).

Reply:
I updated the design. I create the schemas for all the overloaded version of the same operator and then call IsCompatible on all of them to ensure the kernel definition matches the schema. I left some verifications to make sure the kernel definition can be built.